libraries/Microsoft.Bot.Builder.Prompts/OAuthPrompt.cs

```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Adapters;
+using Microsoft.Bot.Builder.Core.Extensions;
+using Microsoft.Bot.Schema;
+using static Microsoft.Bot.Builder.Prompts.PromptValidatorEx;
+
+namespace Microsoft.Bot.Builder.Prompts
+{
+    public class OAuthPromptSettings
+    {
+        public string ConnectionName { get; set; }
+        public string Title { get; set; }
+        public string Text { get; set; }
+    }
+
+    public class TokenResult : PromptResult
+    {
+        public TokenResult() { }
+
+        public TokenResponse Value { get; set; }
+    }
+
+    public class OAuthPrompt
+    {
+        private Regex regex = new Regex(@"(\d{6})");
+        private OAuthPromptSettings _settings;
+        private PromptValidator<TokenResult> _promptValidator;
+
+        public OAuthPrompt(OAuthPromptSettings settings, PromptValidator<TokenResult> validator = null)
+        {
+            _settings = settings ?? throw new ArgumentException(nameof(settings));
+            _promptValidator = validator;
+        }
+
+        public async Task Prompt(ITurnContext context, IMessageActivity activity)
+        {
+            var adapter = context.Adapter as BotFrameworkAdapter;
+            if (adapter == null)
+                throw new InvalidOperationException("OAuthPrompt.Prompt(): not supported by the current adapter");
+
+            if (activity.Attachments == null || activity.Attachments.Count == 0)
+                throw new InvalidOperationException("OAuthPrompt.Prompt(): length of attachments cannot be null");
+
+            var cards = activity.Attachments.Where(a => a.Content is OAuthCard);
```

Comment:
Linq has an "OfType()" method that would be better here.

Reply:
why so? isnt oftype meant for filtering an ienumerable, while i am trying to filter an ienumerable here based on teh type of a property of the elements?

---
In reply to: [184820265](https://github.com/Microsoft/botbuilder-dotnet/pull/520#discussion_r184820265) [](ancestors = 184820265)